File: part2/HaversineCalc/HaversineFormula.cs
using System;
using static System.Math;

public static class HaversineFormula
{

    public static double Square(double a)
    {
        return a * a;
    }

    public static double Deg2Rad(double a)
    {
        return a * Math.PI * 2L / 360L;
    }

    public static double Calc(double x0, double y0, double x1, double y1, double rEarth)
    {
        // i mean...do i have to do this?
        // won't all of the trig functions wrap
        if (x0 < 0) x0 += 360;
        else if (x0 > 180) x0 -= 360;
        if (x1 < 0) x1 += 360;
        else if (x1 > 180) x1 -= 360;
        if (y0 < 0) y1 += 180;
        else if (y0 > 90) y0 -= 180;
        if (y1 < 0) y1 += 180;
        else if (y1 > 180) y1 -= 180;

        double lon1 = x0;
        double lat1 = y0;
        double lon2 = x1;
        double lat2 = y1;

        double dLat = Deg2Rad(lat2 - lat1);
        double dLon = Deg2Rad(lon2 - lon1);
        lat1 = Deg2Rad(lat1);
        lat2 = Deg2Rad(lat2);

        double a = Square(Sin(dLat/2L)) + Cos(lat1) * Cos(lat2) * Square(Sin(dLon/2L));
        double c = 2L * Asin(Sqrt(a));

        double result = rEarth * c;
        return result; 
    }


}

File: part2/cpp_haversine/haversine_formula.cpp
/* This is all just copied from casey muratori's shit
 * i guess we'll have to define the trig functions somewhere
 *  and the f64
 */


static f64 Square(f64 A) 
{
    f64 Result = A*A;
    return Result;
}

static f64 Deg2Rad(f64 Degrees)
{
    f64 Result = 0.01745329251994329577 * Degrees;
    return Result;
}

// casey says the earth's radius is 6472.8
static f64 ReferenceHaversine(f64 X0, f64 Y0, f64 X1, f64 Y1, f64 R_Earth)
{
    f64 lat1 = Y0;
    f64 lat2 = Y1;
    f64 lon1 = X0;
    f64 lon2 = X1;

    f64 dLat = Deg2Rad(lat2 - lat1);
    f64 dLon = Deg2Rad(lon2 - lon1);
    lat1 = Deg2Rad(lat1);
    lat2 = Deg2Rad(lat2);

    f64 a = Square(sin(dLat/2.0)) + cos(lat1)*cos(lat2)*Square(sin(dLon)/2);
    f64 c = 2.0*asin(sqrt(a));

    f64 Result = R_Earth * c;

    return Result;
}

